### DISCUSSION TITLE
pathPrefix & dev mode

### DISCUSSION THREAD

**Discussion by TigersWay** [11ty/eleventy#2417] [Q&A]
Hi,
For the very first time, I'm trying to use `pathPrefix` as it's supposed to: _your site lives in a different subdirectory_
- under Windows 10
- dev mode (--serve)
- with the last canary version & eleventy-dev-server

While I managed to find another way and use other servers already installed; I have a hard time to guess how I am supposed to set it up...

Am I just lacking coffee? Anyone could shed a light?
Thanks!

**Comment by pdehaan** [comment]
I've only used it if I'm planning on deploying to something like GitHub pages and the site would live in a subdirectory, like https://pdehaan.github.io/foo, and I need to prefix all my URLs with "/foo". And then I think the trick is to wrap all your URLs with the `url` filter, which should automatically prepend the `pathPrefix` (which you can specify in your .eleventy.js config or via the command line).

**Comment by TigersWay** [comment]
@pdehaan So far I've used it quite some time to generate multi language sites, the multi-pass way 😄 
But yes, I know or at least understand why to use it (including the url filter)..... But not yet **how**!

The dev server open `localhost/<pathprefix>`. How then am I suppose the set `outputDir`?

**Reply by pdehaan** [reply]
Maybe I'm still missing your question...

I have this .eleventy.js config file:

```js
module.exports = function (eleventyConfig) {

  return {
    pathPrefix: "myprefix",
    dir: {
      input: "src",
      output: "www",
    }
  };
};
```

And this is my src/pages/one.njk file:
```njk
---
title: ONE
---

<h1>{{ title }}</h1>

{{ "/pages/two" | url }}
{{ "/pages/three" | url }}
```

If I build the site, my pages are written to ./www/pages/one/index.html and has the following content:
```html

<h1>ONE</h1>

/myprefix/pages/two
/myprefix/pages/three
```

So you can see that the URLs are prefixed with my specified `pathPrefix` from the .eleventy.js file (or you can set it via the command line using <kbd>eleventy --pathprefix=myprefix</kbd>, but the files will be written to your `dir.output` directory. Setting the `pathPrefix` won't change your output file structure, just the way the URLs are written in your templates (if I understand correctly).
See https://www.11ty.dev/docs/config/#deploy-to-a-subdirectory-with-a-path-prefix and https://www.11ty.dev/docs/filters/url/.

**Comment by TigersWay** [comment]
> Maybe I'm still missing your question...

I have to be honest with you, I'm not sure I'm not totally silly! 🤣 But I believe I understand pathprefix parameter & url filter....

Manwhile when I run something like this:
`npx eleventy --serve --output=www --pathprefix=SUBDIR`
After a few known lines I get this:
`[11ty] Server at http://localhost:8080/SUBDIR`

And of course it doesn't work. I surely have to setup something, but what?
I managed to make it work with some junction and apache, but how/what am I suppose to do with eleventy-dev-server?

**Reply by pdehaan** [reply]
Ah, I was using the latest stable (1.0.1) and I generally don't use `--serve`.

But interestingly, if I create a src/index.njk file and then run run <kbd>eleventy --serve</kbd>, I get the following output:

```
[11ty] Using pathPrefix: myprefix
[11ty] Wrote 4 files in 0.33 seconds (v1.0.1)
[11ty] Watching…
[Browsersync] Access URLs:
 -----------------------------------
    Local: http://localhost:8080
```

But if I go to http://localhost:8080, it redirects to http://localhost:8080/myprefix/ and my index.html file correctly renders.
Let me try updating to 2.0.0-canary and see if that gives different results.

**Reply by pdehaan** [reply]
Yeah, might be a Canary issue:

```
[11ty] Wrote 4 files in 0.04 seconds (v2.0.0-canary.11)
[11ty] Watching…
[11ty] Server at http://localhost:8081/myprefix
```

If I try that link, it seems to throw errors.
But if I go to http://localhost:8080/myprefix/ with a trailing slash it seems to work.

Although, I don't have the "proper" `<html>` and `<head></head><body></body>` tags, so maybe something is happening there.
But I did verify in my output www/ folder and all the files are there like I expected.

**Comment by TigersWay** [comment]
@pdehaan You are 100% right! And I did not even think about trying to add the trailing slash. now `http://localhost:8080/SUBDIR/` is working fine!!!!

**Reply by TigersWay** [reply]
Heading to create a little issue then

**Reply by pdehaan** [reply]
For future travellers, the issue in question is https://github.com/11ty/eleventy/issues/2418

**Comment by zachleat** [comment]
A few fixes going into dev-server 1.0.0-canary.14. See https://github.com/11ty/eleventy-dev-server/issues/37#issuecomment-1211321097
